sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -649,13 +664,16 @@ object CollapseProject extends Rule[LogicalPlan] {
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
     case p1 @ Project(_, p2: Project) =>
-      if (haveCommonNonDeterministicOutput(p1.projectList, p2.projectList)) {
+      if (haveCommonNonDeterministicOutput(p1.projectList, p2.projectList) ||
+        ScalarSubquery.hasScalarSubquery(p1.projectList) ||
+        ScalarSubquery.hasScalarSubquery(p2.projectList)) {
```

Comment:
why do we allow it before?

Reply:
@cloud-fan.. Let me get back to you on this, need to debug again :-)

Reply:
@cloud-fan One failing test that i needed to address with this change is in subquerysuite.
```
select a, (select sum(b) from l l2 where l2.a <=> l1.a) sum_b from l l1")
```
One main reason is, the Filter ops with outer references were pulled up before optimizeSubqueries rule. So by the time other optimization rules kick in (like pushDownPredicate etc), it does not see outer references. But with the change in the PR, they are present. So another way to handle this is to change pushdownPredicate rule to make sure the filter clauses with outer references are not moved down. May be thats better way to handle it and keep CollapseProject as it is.